File: ios/ZebraFactory.swift
//
//  ZebraFactory.swift
//  ZebraStriped
//
//  Created by Dawid Zawada on 25/04/2024.
//

import ZXingObjC

struct ZebraFactory {
    
    static let DEFAULT_ON_COLOR = UIColor.black.cgColor
    static let DEFAULT_OFF_COLOR = UIColor.white.cgColor
    
    static func generateCodeImage(value: String, format: ZXBarcodeFormat, size: CGSize, onColor _onColor: CGColor?, offColor _offColor: CGColor?) -> UIImage {
        let writer = ZXMultiFormatWriter();
        let hints = ZXEncodeHints()
        hints.margin = 0
        
        do {
            let result = try writer.encode(value, format: format, width: Int32(size.width), height: Int32(size.height), hints: hints)
            let onColor = _onColor ?? DEFAULT_ON_COLOR
            let offColor = _offColor ?? DEFAULT_OFF_COLOR
            return UIImage(cgImage: ZXImage(matrix: result, on: onColor, offColor: offColor).cgimage)
        } catch {
            return UIImage()
        }
        
    }
    
    static func parseToBarcodeFormat(codeFormat: CodeFormat) -> ZXBarcodeFormat {
        switch codeFormat {
        case .ean8:
            return kBarcodeFormatEan8
        case .ean13:
            return kBarcodeFormatEan13
        case .upcA:
            return kBarcodeFormatUPCA
        case .upcE:
            return kBarcodeFormatUPCE
        case .code39:
            return kBarcodeFormatCode39
        case .code93:
            return kBarcodeFormatCode93
        case .code128:
            return kBarcodeFormatCode128
        case .itf:
            return kBarcodeFormatITF
        case .codabar:
            return kBarcodeFormatCodabar
        case .qrcode:
            return kBarcodeFormatQRCode
        case .pdf417:
            return kBarcodeFormatPDF417
        case .dataMatrix:
            return kBarcodeFormatDataMatrix
        case .aztec:
            return kBarcodeFormatAztec
        }
    }
    
}

File: android/src/main/java/expo/modules/zebrastriped/ZebraFactory.kt
package expo.modules.zebrastriped

import android.graphics.Bitmap
import android.graphics.Color
import com.google.zxing.BarcodeFormat
import com.google.zxing.EncodeHintType
import com.google.zxing.MultiFormatWriter
import net.glxn.qrgen.android.MatrixToImageConfig
import net.glxn.qrgen.android.MatrixToImageWriter
import java.lang.Exception

class ZebraFactory {

    val DEFAULT_ON_COLOR = Color.BLACK
    val DEFAULT_OFF_COLOR = Color.WHITE

    fun generateCodeBitmap(value: String, format: BarcodeFormat, size: Pair<Int, Int>, onColor: String?, offColor: String?): Bitmap? {

        val (width, height) = size

        if(value.isEmpty()){
            return null
        }

        val writer = MultiFormatWriter()

        val hints: MutableMap<EncodeHintType, Any> = mutableMapOf(
                EncodeHintType.MARGIN to 0
        )

        val matrix = writer.encode(value, format, width, height, hints)
        val _onColor = onColor?.let { Color.parseColor(it) } ?: DEFAULT_ON_COLOR
        val _offColor = offColor?.let { Color.parseColor(it) } ?: DEFAULT_OFF_COLOR


        return MatrixToImageWriter.toBitmap(matrix, MatrixToImageConfig(_onColor, _offColor))
    }

    fun parseToBarcodeFormat(codeFormat: CodeFormat): BarcodeFormat {
        val barcodeFormat = when (codeFormat) {
            CodeFormat.EAN8 -> BarcodeFormat.EAN_8
            CodeFormat.EAN13 -> BarcodeFormat.EAN_13
            CodeFormat.UPCA -> BarcodeFormat.UPC_A
            CodeFormat.UPCE -> BarcodeFormat.UPC_E
            CodeFormat.CODE39 -> BarcodeFormat.CODE_39
            CodeFormat.CODE93 -> BarcodeFormat.CODE_93
            CodeFormat.CODE128 -> BarcodeFormat.CODE_128
            CodeFormat.ITF -> BarcodeFormat.ITF
            CodeFormat.CODABAR -> BarcodeFormat.CODABAR
            CodeFormat.QRCODE -> BarcodeFormat.QR_CODE
            CodeFormat.PDF417 -> BarcodeFormat.PDF_417
            CodeFormat.DATAMATRIX -> BarcodeFormat.DATA_MATRIX
            CodeFormat.AZTEC -> BarcodeFormat.AZTEC
        }

        return barcodeFormat
    }
}